x-pack/legacy/plugins/apm/typings/elasticsearch/index.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Omit } from 'utility-types';
+import { SearchParams, SearchResponse } from 'elasticsearch';
+import { AggregationResponseMap, AggregationInputMap } from './aggregations';
+import { StringMap } from '../common';
+
+export interface ESSearchBody {
+  query?: any;
+  size?: number;
+  aggs?: AggregationInputMap;
+  track_total_hits?: boolean | number;
+}
+
+export type ESSearchRequest = Omit<SearchParams, 'body'> & {
+  body?: ESSearchBody;
+};
+
+export interface ESSearchOptions {
+  restTotalHitsAsInt: boolean;
+}
+
+export type ESSearchHit<T> = SearchResponse<T>['hits']['hits'][0];
+
+export type ESSearchResponse<
+  TDocument,
+  TSearchRequest extends ESSearchRequest,
+  TOptions extends ESSearchOptions = { restTotalHitsAsInt: false }
+> = Omit<SearchResponse<TDocument>, 'aggregations' | 'hits'> &
+  (TSearchRequest extends { body: { aggs: AggregationInputMap } }
+    ? {
+        aggregations?: AggregationResponseMap<TSearchRequest['body']['aggs']>;
+      }
+    : {}) &
+  ({
+    hits: Omit<SearchResponse<TDocument>['hits'], 'total'> &
+      (TOptions['restTotalHitsAsInt'] extends true
```

Comment:
How does this work? Can it pass the options, and if `restTotalHitsAsInt` is `true` it will type the response accordingly? That's pretty cool 👍 

One thing: Mostly `restTotalHitsAsInt` is written as `rest_total_hits_as_int` so many we should require it as such?

Reply:
`restTotalHitsAsInt` is only used by Lens:
![image](https://user-images.githubusercontent.com/209966/66931391-4cdd9100-f036-11e9-8442-6f079fda2bb4.png)

`rest_total_hits_as_int` is used 54 places
![image](https://user-images.githubusercontent.com/209966/66931415-5666f900-f036-11e9-897e-4a03d325bb83.png)

Reply:
I figured it would look kind of weird if we add more options. If we add something, for some reason, that does not have a query parameter equivalent, do we define it in snake_case then as well?

Reply:
The bigger issue here is not snake case, but it's that this type is really confusing in practice. I tried getting the elasticsearch-js client to improve typings in the way you're proposing here, but there are some strong caveats.

Here is a discussion I had with @delvedor: https://github.com/elastic/elasticsearch-js/pull/970

Specifically, I've extended the playground example [here to demonstrate why this is is so tricky](https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgEJwM4QMoTlBACwCUIMAHAexC2QG8AoZZ5OAczagjbjGGowB+AFzIAriADWISgHcQAbgYBfBgzABPcilz4ipCgIgAeUgEcxZMMgC84qTPkA+W8nOWM1iAA9IIACYY9MhcngD6YJRgcAA2YYTAYBhhmGGgYKJgUJbIysiCTCzoWLoEJGRUNCgAZPSFLCwJSaIgYgC2AEbQ9bnIosU4eGUGlbS1jA0NTRiidMgAbrGWLe1dUAohEDG8-CCiAORskPvIAD7I+xBm+6oNykpqCALWoaMQAHKr0KKl+hVGxjmoTAESisXiiWSqXSmWyKGULjsE0akNEAFYADQ9dicbg7ASzZRY+6PZ6bQxVADyHQAVhAEBlkL9yhSsIDySDItE4tMUskYch4DFaAjXMjmNNZgslhB0RjNts+NQDlcTkTsRwuDwlTRCcSHk8aC9-lTafSwAAmH5DP6slBInqS4KLGLLZCYhX4vYXVW5LENHFar0zejqkkMQ2eTYWKzUukMsUc0HciFJPlpcCwnLhyPGu1x80AZmtehZb2Mmm0lBg0Y8YALDMRdSmqOdMrlnp1Kuufo1uO1uxDdDDSiAA)

The biggest caveat here is that `Response<{ rest_total_hits_as_int: true}>` is not the same as ```
const request = { rest_total_hits_as_int: true };
Response<typeof request>;
```

For example, this type is totally valid, which I found unintuitive:

```
const f = { result: true };
const r: typeof f['result'] = false;
```

Reply:
I think TS automatically widens boolean and string literals to respectively booleans and strings. If you change that to `{ result: true as const }`, it will show a TS error.

A type definition (rather than a type inferred from a value) does not have the same issue. `true` remains `true`, rather than being widened to `boolean`.

Reply:
Right, that's what I meant by "confusing in practice." That being said, this is optional and has the correct defaults. So I think this is fine.